h2/src/main/org/h2/util/DateTimeUtils.java

```diff
@@ -1497,4 +1917,313 @@ public static long convertScale(long nanosOfDay, int scale) {
+            return intervalFromAbsolute(qualifier, absolute, 1_000_000_000);
+        case YEAR_TO_MONTH:
+            return intervalFromAbsolute(qualifier, absolute, 12);
+        case DAY_TO_HOUR:
+            return intervalFromAbsolute(qualifier, absolute.divide(BigInteger.valueOf(3_600_000_000_000L)), 24);
+        case DAY_TO_MINUTE:
+            return intervalFromAbsolute(qualifier, absolute.divide(BigInteger.valueOf(60_000_000_000L)), 24 * 60);
+        case DAY_TO_SECOND:
+            return intervalFromAbsolute(qualifier, absolute, NANOS_PER_DAY);
+        case HOUR_TO_MINUTE:
+            return intervalFromAbsolute(qualifier, absolute.divide(BigInteger.valueOf(60_000_000_000L)), 60);
+        case HOUR_TO_SECOND:
+            return intervalFromAbsolute(qualifier, absolute, 3_600_000_000_000L);
+        case MINUTE_TO_SECOND:
+            return intervalFromAbsolute(qualifier, absolute, 60_000_000_000L);
+        default:
+            throw new IllegalArgumentException();
+        }
+    }
+
+    private static ValueInterval intervalFromAbsolute(IntervalQualifier qualifier, BigInteger absolute, long divisor) {
+        BigInteger[] dr = absolute.divideAndRemainder(BigInteger.valueOf(divisor));
+        return ValueInterval.from(qualifier, absolute.signum() < 0, leadingExact(dr[0]), Math.abs(dr[1].longValue()));
+    }
+
+    private static long leadingExact(BigInteger absolute) {
+        if (absolute.compareTo(BigInteger.valueOf(999_999_999_999_999_999L)) > 0
+                || absolute.compareTo(BigInteger.valueOf(-999_999_999_999_999_999L)) < 0) {
+            throw DbException.get(ErrorCode.NUMERIC_VALUE_OUT_OF_RANGE_1, absolute.toString());
+        }
+        return Math.abs(absolute.longValue());
+    }
+
+    /**
+     * @param qualifier
+     *            qualifier
+     * @param negative
+     *            whether interval is negative
+     * @param leading
+     *            value of leading field
+     * @param remaining
+     *            values of all remaining fields
+     * @return years, or 0
+     */
+    public static long yearsFromInterval(IntervalQualifier qualifier, boolean negative, long leading, long remaining) {
+        if (qualifier == IntervalQualifier.YEAR || qualifier == IntervalQualifier.YEAR_TO_MONTH) {
+            long v = leading;
+            if (negative) {
+                v = -v;
+            }
+            return v;
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * @param qualifier
+     *            qualifier
+     * @param negative
+     *            whether interval is negative
+     * @param leading
+     *            value of leading field
+     * @param remaining
+     *            values of all remaining fields
+     * @return months, or 0
+     */
+    public static long monthsFromInterval(IntervalQualifier qualifier, boolean negative, long leading, long remaining)
+    {
+        long v;
+        if (qualifier == IntervalQualifier.MONTH) {
```

Comment:
do we need to handle IntervalQualifier.YEAR here? Just asking, didn't check the callsites

Reply:
No, default 0 is a correct answer.